src/syntaxtree.c

```diff
@@ -31,50 +31,92 @@
 // assuming that the set of their respective ids are disjoint.
 static const CompilerT Compilers[];
 #define COMPILER_ARITY(...) (sizeof((ArgT[]){ __VA_ARGS__ }) / sizeof(ArgT))
-#define COMPILER_GENERIC(tnum, compiler, ...)                                \
+#define COMPILER_GENERIC(tnum, compiler, coder, ...)                         \
     [tnum] = {                                                               \
-        tnum, compiler, #tnum, COMPILER_ARITY(__VA_ARGS__), { __VA_ARGS__ }  \
+        tnum,           compiler, coder, #tnum, COMPILER_ARITY(__VA_ARGS__), \
+        { __VA_ARGS__ }                                                      \
     }
 
-#define COMPILER(tnum, compiler) COMPILER_GENERIC(tnum, compiler)
+#define COMPILER(tnum, compiler, coder)                                      \
+    COMPILER_GENERIC(tnum, compiler, coder)
 
 #define COMPILER_(tnum, ...)                                                 \
-    COMPILER_GENERIC(tnum, SyntaxTreeDefaultCompiler, __VA_ARGS__)
+    COMPILER_GENERIC(tnum, SyntaxTreeDefaultCompiler,                        \
+                     SyntaxTreeDefaultCoder, __VA_ARGS__)
 
-#define ARG(name, func)                                                      \
+#define ARG(name, func, coder)                                               \
     {                                                                        \
-        name, func                                                           \
+        name, func, coder                                                    \
     }
-#define ARG_(name) ARG(name, SyntaxTreeCompiler)
+#define ARG_(name) ARG(name, SyntaxTreeCompiler, SyntaxTreeDefaultCoder)
 
-#define ARGS(name) ARG(name, 0)
+#define ARGS(name) ARG(name, 0, 0)
 
 
-static Obj typeStrings;
+static Obj typeStrings;    // plain list mapping statement tnums to strings
+                           // like "T_INTEXPR"
+static Obj typeRec;    // plain record mapping strings like "T_INTEXPR" to the
+                       // corresponding statement tnum
 
+static Obj ElmRecST(UInt1 tnum, Obj node, const char * name)
+{
+    UInt rnam = RNamName(name);
+    if (!IsbPRec(node, rnam)) {
+        const char * typename = Compilers[tnum].name;
+        ErrorQuit("while coding %s: node field <%s> must be present",
+                  (Int) typename, (Int)name);
+    }
+    return ElmPRec(node, rnam);
+}
 
 static Obj SyntaxTreeFunc(Obj result, Obj func);
 
+static UInt1 FIND_TNUM_FROM_STRING(Obj typestr)
+{
+    UInt rnam = RNamObj(typestr);
+    if (IsbPRec(typeRec, rnam)) {
+        return (UInt1)UInt_ObjInt(ElmPRec(typeRec, rnam));
+    }
+    ErrorQuit("Unrecognized syntax tree node type %s",
+              (Int)CSTR_STRING(typestr), 0);
+}
+
+static UInt1 GetTypeTNum(Obj node)
+{
+    UInt type = RNamName("type");
+    if (!IsbPRec(node, type)) {
+        ErrorQuit("while coding: <type> field is not present", 0, 0);
+    }
+    return FIND_TNUM_FROM_STRING(ELM_REC(node, type));
```

Comment:
```suggestion
    return FIND_TNUM_FROM_STRING(ElmPRec(node, type));
```
You just called `IsbPRec` a few lines before, and use `ElmPRec`, so might as well use that here. Or, if we truly want to support general records (in case somebody ever adds some), we should be using `ISB_REC`, `UNB_REC`, `ELM_REC` everywhere. But I think it's OK to restrict to precs